lms/djangoapps/discussion_api/api.py

```diff
@@ -177,103 +185,136 @@ def get_course(request, course_key):
+                discussion_topic = DiscussionTopic(
+                    module.discussion_id,
+                    module.discussion_target,
+                    get_thread_list_url(request, course_key, [module.discussion_id]),
+                )
+                children.append(discussion_topic.__dict__)
+
+                if topic_ids and module.discussion_id in topic_ids:
+                    existing_topic_ids.append(module.discussion_id)
+
+        if not topic_ids or children:
+            discussion_topic = DiscussionTopic(
+                None,
+                category,
+                get_thread_list_url(request, course_key, [item.discussion_id for item in get_sorted_modules(category)]),
+                children,
+            )
+            courseware_topics.append(discussion_topic.__dict__)
 
-    courseware_topics = [
-        {
-            "id": None,
-            "name": category,
-            "thread_list_url": get_thread_list_url(
-                request,
-                course_key,
-                [item.discussion_id for item in get_sorted_modules(category)]
-            ),
-            "children": [
-                {
-                    "id": module.discussion_id,
-                    "name": module.discussion_target,
-                    "thread_list_url": get_thread_list_url(request, course_key, [module.discussion_id]),
-                    "children": [],
-                }
-                for module in get_sorted_modules(category)
-            ],
-        }
-        for category in sorted(modules_by_category.keys())
-    ]
-
-    non_courseware_topics = [
-        {
-            "id": entry["id"],
-            "name": name,
-            "thread_list_url": get_thread_list_url(request, course_key, [entry["id"]]),
-            "children": [],
-        }
-        for name, entry in sorted(
-            course.discussion_topics.items(),
-            key=lambda item: item[1].get("sort_key", item[0])
-        )
-    ]
+    return courseware_topics, existing_topic_ids
 
-    return {
-        "courseware_topics": courseware_topics,
-        "non_courseware_topics": non_courseware_topics,
-    }
 
+def get_non_courseware_topics(request, course_key, course, topic_ids):
+    """
+    Returns a list of topic trees that are not linked to courseware.
+
+    Parameters:
+
+        request: The django request objects used for build_absolute_uri.
+        course_key: The key of the course to get discussion threads for.
+        course: The course for which topics are requested.
+        topic_ids: A list of topic IDs for which details are requested.
+            This is optional. If None then all course topics are returned.
+
+    Returns:
+        A list of non-courseware topics and a list of existing topics among
+        topic_ids.
 
-def get_course_topic(request, course_key, topic_id):
     """
-    Returns details of discussion topic for the given topic_id.
+    non_courseware_topics = []
+    existing_topic_ids = []
```

Comment:
`existing_topic_ids` is also used as a set when it gets returned.  Make it a set here, and save the conversion later.